### PR TITLE
vi - branches

### PR DESCRIPTION
**Pull request by criacuervos**
# Task List
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe in your own words what the Model is doing in Rails | The model in my case is handling some of the logic of what constitutes a complete and incomplete task. In a more abstract sense, it communicates with the database, in which data passes through it before it heads to the controller or vice versa. 
Describe in your own words what the Controller is doing in Rails | The controller is the program mini manager! It communicates with the web server,  and decides what to do when handed information from the model or the view. in charge of defining different actions, which is how it knows what to send back.
Describe in your own words what the View is doing in Rails | The view is basically rendered data that is nice to look at.  We can use the view files to transform data with HTML. It knows what a controllers instance variables are.
Describe an edge-case controller test you wrote | If someone were to try to delete something twice (unlikely to happen, but still a good test)
What is the purpose of using strong params? (i.e. the params method in the controller) | Mainly for security, so that some stranger can't make an API request and blow up my site with all sorts of requests. The parameters that we specifically want are hidden.
How are Rails migrations related to Rails models? | While a model and first migration is what gives us our initial table of data, the way in which they continue to be related are that migrations are used whenever modifications are made to that table's contents.
Describe one area of Rails that are still unclear on | I'm definitely still fuzzy on where stuff goes. While I like to think I understand the broader picture, the nitty gritty details of how everything is interrelated and what that specifically means for where certain logic should go is still confusing. For example, I wasn't sure where I should put the logic for updating a completed task's time to the current time when you check mark the box. (It's the controller, which sort of makes sense now I think?). Also writing tests for our custom actions.

### CODE REVIEW

**Comment by tildeee** [comment]
# Task List
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Baseline** |
Appropriate Git Usage with no extraneous files checked in | x
Answered comprehension questions | x
Successfully handles: Index, Show | x
Index & Show tests pass | x
Successfully handles: New, Create | x
New & Create tests pass | x
Successfully handles: Edit, Update | x
Tests for Edit & Update test valid & invalid task ids | x
Successfully handles: Destroy, Task Complete | x
Tests for Destroy & Task Complete include tests for valid and invalid task ids | x
Routes follow RESTful conventions | x
Uses named routes (like `_path`) | x
**Overall** |



Fantastic work on this project, Vi!! Your controller code is very clean, and your tests are thorough, clean, and have great form and style! Also, super well done on the "mark complete" feature-- the multiple select of the tasks is really really cool.


I have a few comments on concrete ways to make your code more Rails-y, but otherwise I don't have too many comments; the project works great!


Well done! Keep it up!

File: app/controllers/tasks_controller.rb [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    if @task.nil?
+      redirect_to root_path
+      return
+    end 
+    
+    @task.name = params[:task][:name]
+    @task.description = params[:task][:description]
+    @task.completed = params[:task][:completed]
+    
+    if @task.save
```
Comment (tildeee) [inline app/controllers/tasks_controller.rb:56]: In order to take advantage of our strong params `task_params`, we might want to refactor this to `if @task.update(task_params)` and get rid of the `@task.<something> = params[<something>]` above

File: test/controllers/tasks_controller_test.rb [hunk 2]
```diff
@@ -79,48 +82,122 @@
...
+      #Need to change to 'assert_nil' but having trouble finding correct syntax and documentation for this
 
       must_respond_with :redirect
       must_redirect_to task_path(new_task.id)
     end
+
+    it "will raise an error for creating a task with invalid arguments" do 
+      expect { post tasks_path, params: invalid_task_hash }.must_differ "Task.count", 0
+      must_respond_with 200
+      #Weird bug here I think? I want to figure out how to check for a rendered form, but my page redirects to /tasks despite displaying the form again with if invalid entry. On StackOverflow I found out it might be because #index is before #create in routes - but I also can't figure out how to rewrite the order of that.
```
Comment (tildeee) [inline test/controllers/tasks_controller_test.rb:95]: We might have talked about this, but checking for a render is actually really difficult-- in this case, just checking for responding with success is good enough!

File: test/controllers/tasks_controller_test.rb [hunk 3]
```diff
@@ -79,48 +82,122 @@
...
+
+      expect(Task.find_by(id: existing_task.id).completed).must_equal DateTime.new(10/6/19)
+
+      must_redirect_to task_path(existing_task.id)
+
     end
 
     it "will redirect to the root page if given an invalid id" do
-      # Your code here
+      invalid_id = [-1, "woops"]
```
Comment (tildeee) [inline test/controllers/tasks_controller_test.rb:147]: Fantastic test!

File: test/controllers/tasks_controller_test.rb [hunk 4]
```diff
@@ -79,48 +82,122 @@
...
   # Complete for Wave 4
-  describe "toggle_complete" do
-    # Your tests go here
+  describe "completed task" do
+    it "completed changes from nil to current datetime, and redirects to home" do
+      #arrange
+      test_task = Task.create(name: "Testing", description: "in process")
+
+      #act
+      patch completed_task(id:test_task.id), params: {destination: "root"}
```
Comment (tildeee) [inline test/controllers/tasks_controller_test.rb:197]: In order to get this to work, you'll need to do `put completed_tasks_path(id:test_task.id), params: {destination: "root"}`
